pkg/controllers/coscheduling/podgroup_controller.go

```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
+		return ctrl.Result{}, nil
+	}
+
+	pgCopy := pg.DeepCopy()
+	podList := &corev1.PodList{}
+	err = r.List(ctx, podList,
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(map[string]string{
+				schedv1alpha1.PodGroupLabel: pgCopy.Name}).AsSelector(),
+		})
+	if err != nil {
+		log.Error(err, "List pods for group failed")
+		return ctrl.Result{}, err
+	}
+	pods := podList.Items
+
+	switch pgCopy.Status.Phase {
+	case "":
+		pgCopy.Status.Phase = schedv1alpha1.PodGroupPending
+	case schedv1alpha1.PodGroupPending:
+		if len(pods) >= int(pg.Spec.MinMember) {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupPreScheduling
+			fillOccupiedObj(pgCopy, &pods[0])
+		}
+	default:
+		var (
+			running   int32 = 0
+			succeeded int32 = 0
+			failed    int32 = 0
+		)
+		if len(pods) != 0 {
+			for _, pod := range pods {
+				switch pod.Status.Phase {
+				case v1.PodRunning:
+					running++
+				case v1.PodSucceeded:
+					succeeded++
+				case v1.PodFailed:
+					failed++
+				}
+			}
+		}
+		pgCopy.Status.Failed = failed
+		pgCopy.Status.Succeeded = succeeded
+		pgCopy.Status.Running = running
 
-	// TODO(user): your logic here
+		if len(pods) == 0 {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupPending
+			break
+		}
 
-	return ctrl.Result{}, nil
+		if pgCopy.Status.Scheduled >= pgCopy.Spec.MinMember && pgCopy.Status.Phase == schedv1alpha1.PodGroupScheduling {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupScheduled
+		}
+
+		if pgCopy.Status.Succeeded+pgCopy.Status.Running >= pg.Spec.MinMember && pgCopy.Status.Phase == schedv1alpha1.PodGroupScheduled {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupRunning
+		}
+		// Final state of pod group
+		if pgCopy.Status.Failed != 0 && pgCopy.Status.Failed+pgCopy.Status.Running+pgCopy.Status.Succeeded >= pg.Spec.
+			MinMember {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupFailed
+		}
+		if pgCopy.Status.Succeeded >= pg.Spec.MinMember {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupFinished
+		}
+	}
+
+	err = r.patchPodGroup(ctx, pg, pgCopy)
+	return ctrl.Result{}, err
+}
+
+func (r *PodGroupReconciler) patchPodGroup(ctx context.Context, old, new *schedv1alpha1.PodGroup) error {
+	if !reflect.DeepEqual(old, new) {
+		patch := client.MergeFrom(old)
+
+		err := r.Patch(ctx, new, patch)
+		if err != nil {
+			return err
+		}
+		err = r.Status().Patch(ctx, new, patch)
+		return err
+	}
+	return nil
+}
+
+func fillOccupiedObj(pg *schedv1alpha1.PodGroup, pod *corev1.Pod) {
+	var refs []string
+	for _, ownerRef := range pod.OwnerReferences {
+		refs = append(refs, fmt.Sprintf("%s/%s", pod.Namespace, ownerRef.Name))
+	}
+	if len(pg.Status.OccupiedBy) == 0 {
+		return
+	}
+	if len(refs) != 0 {
+		sort.Strings(refs)
+		pg.Status.OccupiedBy = strings.Join(refs, ",")
+	}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodGroupReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.recorder = mgr.GetEventRecorderFor("PodGroupController")
```

Comment:
this recorder is unused you can just not initiate it.

Reply:
it will be used later, initiating this when first migrated to controller runtime and we can use it directly if need to

Reply:
Than you can add it later, when you will use it.